### PR TITLE
upload all built packages, except for the excluded ones

### PR DESCRIPTION
**Pull request by CJ-Wright**
CC: @isuruf @beckermr

### CODE REVIEW

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py [hunk 1]
```diff
@@ -120,7 +122,12 @@ def upload_or_check(recipe_dir, owner, channel, variant):
     # The list of built/not skipped distributions
     built_distributions = [(m, path)
                            for m, _, _ in metas
-                           for path in conda_build.api.get_output_file_paths(m)
+                           # TODO: flip this over to .conda when that format
+                           #  is in flight
+                           for path in glob.glob(os.path.join(
+            conda_build.config.croot,
+            'noarch' if m.noarch or m.noarch_python else m.config.host_subdir,
+            '*.tar.bz2'))
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/upload_or_check_non_existence.py]: For each output name, you are trying to upload all files.

Comment (CJ-Wright) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py]: Can you expand on that some? I don't understand if that is what I'm trying to do and doing it incorrectly, or if that is what I should be doing and I'm not

Comment (isuruf) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py]: What this code did is for each output, calculate the build numbers and the full name(there might be many in rare cases). What the new code is doing is for each output, get the full name of all the outputs.

I think we should get rid of the conda build rendering here and just look at all the packages in `os.path.join(conda_build.config.croot, conda_build.config.subdir)` and `os.path.join(conda_build.config.croot, 'noarch')`.

One issue is that `built_distribution_already_exists` expects a conda-build meta object, but it is used for getting the name and version of the package which can be done by parsing the name as in https://github.com/conda-forge/cf-mark-broken/blob/master/mark_broken.py#L6-L13

Comment (CJ-Wright) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py]: how do you want to handle the `m.skip` logic? That depends on the rendered meta.yaml

Comment (beckermr) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py]: Why do we need skip logic at all? Shouldn't conda-build have not produced packages for things that are skipped?

**Comment by CJ-Wright** [comment]
@isuruf does this implement what you had in mind?

**Comment by CJ-Wright** [comment]
They don't have the needed [`.tar.bz2` extension](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/72/files#diff-2c55b5e34665621d4e425217f0f28912R120). The repodata files are either `.json` or `.json.bz2`

**Comment by CJ-Wright** [comment]
I haven't checked it. I might be able to check it but not soon.

**Comment by CJ-Wright** [comment]
@isuruf what is the best way to check this?

**Comment by isuruf** [comment]
Copy the `upload_or_check_non_existence.py` to the recipe folder of a test feedstock and rerender. This used to work with conda-smithy, but not sure now.

**Comment by isuruf** [comment]
Changing the line https://github.com/conda-forge/conda-smithy/blob/master/conda_smithy/configure_feedstock.py#L683 to `if platform == "linux"` should fix the rerendering.

**Comment by CJ-Wright** [comment]
It worked! 

https://github.com/conda-forge/cf-autotick-bot-test-package-feedstock/pull/40
https://anaconda.org/conda-forge/cf-autotick-bot-test-package/files

**Comment by isuruf** [comment]
@CJ-Wright, you need a rerender

**Comment by CJ-Wright** [comment]
I did a rerender, it said there was nothing to do.

**Comment by CJ-Wright** [comment]
Let me try again

**Comment by isuruf** [comment]
It looks like this functionality was accidentally removed from conda-smithy.

**Comment by CJ-Wright** [comment]
Yep I tried again, no changes in the rerender

**Comment by CJ-Wright** [comment]
what is the best path forward here?

**Comment by beckermr** [comment]
I think we'd need to release a dev version of this package on a dev channel, and then hack the CI scripts in another feedstock to pull it. I think someone rewrote this to install a real python package with command-line tools and so conda-smithy cannot override them.

**Comment by isuruf** [comment]
We can change the logic at https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/master/recipe/conda_forge_ci_setup/upload_or_check_non_existence.py to use the one from the recipe directory if it exists and use it. We can merge with just that change, and then merge this PR.

**Comment by beckermr** [comment]
Makes sense @isuruf. Note that smithy is using the entry point shipped with this package, so we'll want to make sure to change the logic so the entry point looks to the right thing or w/e.

**Comment by isuruf** [comment]
I wasn't thinking of overriding the entry point, just the `upload_or_check_non_existence.py` file.

**Comment by CJ-Wright** [comment]
My plan is to have the first line of the function see if the file is in the correct dir, if so load that file and then run the function from there.

**Comment by isuruf** [comment]
Sounds good. Thanks for working on this.

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py [hunk 2]
```diff
@@ -89,6 +100,16 @@ def distribution_exists_on_channel(binstar_cli, meta, fname, owner, channel='mai
...
 def upload_or_check(recipe_dir, owner, channel, variant):
+    # If there is a dedicated upload in the recipe use that one
+    if 'upload_or_check_non_existence.py' in os.listdir(recipe_dir):
+        import importlib.util
+        spec = importlib.util.spec_from_file_location('upload_two',
+                                                      os.path.join(recipe_dir,
+                                                                   'upload_or_check_non_existence.py'))
+        mod = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(mod)
+        return mod.upload_or_check(recipe_dir, owner, channel, variant)
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:111]: Wouldn't this go into an endless loop?

Comment (CJ-Wright) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:111]: right let me put a flag in the call.

Comment (CJ-Wright) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:111]: Actually technically not, you just have to *not* have this inspection clause in the recipe version.

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py [hunk 3]
```diff
@@ -89,6 +100,16 @@ def distribution_exists_on_channel(binstar_cli, meta, fname, owner, channel='mai
 
 
 def upload_or_check(recipe_dir, owner, channel, variant):
+    # If there is a dedicated upload in the recipe use that one
+    if 'upload_or_check_non_existence.py' in os.listdir(recipe_dir):
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:104]:
```suggestion
    if os.path.exists(os.path.join(recipe_dir, "conda_forge_ci_setup", "upload_or_check_non_existence.py")):
```

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py [hunk 4]
```diff
@@ -89,6 +100,16 @@ def distribution_exists_on_channel(binstar_cli, meta, fname, owner, channel='mai
...
+        import importlib.util
+        spec = importlib.util.spec_from_file_location('upload_two',
+                                                      os.path.join(recipe_dir,
```
Comment (isuruf) [inline recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:107]:
```suggestion
                                                      os.path.join(recipe_dir, "conda_forge_ci_setup",
```

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py (hunk 2 above)
Comment (isuruf) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:111]: I was thinking of using the new version of this file to upload the new package, so that we don't have to test in a test feedstock. This feedstock will be its own test. That's how it was designed first.

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py (hunk 4 above)
Comment (CJ-Wright) [reply recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:107]: Ah so the plan is to include the entire ci setup folder?

**Comment by isuruf** [comment]
Can you revert back to 38e3ddd9039eeec72377d9295ba78204977add6f and merge with master?

**Comment by beckermr** [comment]
It occurred to me that a more robust way to version the pinnings would be to use the time stamp of the commit. We could change the CI scripts to inject that as an environment variable and then use. 

Using the datetime at build is subject to race conditions on CI jobs and is not reproducible.

**Comment by beckermr** [comment]
Cc @isuruf

**Comment by isuruf** [comment]
In that case, you can add a variable like `GIT_COMMITTER_DATE` to https://github.com/conda/conda-build/blob/a563eb19bdc69552836540a80021ffe3fc417ee4/conda_build/environ.py#L159

**Comment by beckermr** [comment]
Wait no. This would be the git commit hash from the repo of the source, not the feedstock. We could change the pinnings recipe to point at master, but then that would break CI. I think we would need to put this in smithy or the ci scripts here.

**Comment by isuruf** [comment]
I think you can make the recipe be the source. Not sure.

**Comment by beckermr** [comment]
Yup, but that would break PRs since the source would change.

**Comment by beckermr** [comment]
+1 to merge this @isuruf @CJ-Wright 

it doesn't impact the change i'd like to put in specifically for versioning with the commit date

**Comment by CJ-Wright** [comment]
Weird windows failures

File: recipe/conda_forge_ci_setup/upload_or_check_non_existence.py [hunk 5]
```diff
@@ -14,6 +15,17 @@
...
 from conda_build.conda_interface import get_index
 import conda_build.api
+import conda_build.config
+
+
+def split_pkg(pkg):
+    if not pkg.endswith(".tar.bz2"):
+        raise RuntimeError("Can only process packages that end in .tar.bz2")
+    pkg = pkg[:-8]
+    plat, pkg_name = pkg.split("/")
```
Comment (CJ-Wright) [inline recipe/conda_forge_ci_setup/upload_or_check_non_existence.py:25]: This might not work on windows boxes. Most likely should use actual os path tools.

**Comment by CJ-Wright** [comment]
@conda-forge/core can this get a merge? I'd love to get the new pinning version set up so we can merge a bunch of stuff.
